Update numba to 0.56.3

AndrewVallette opened the pull request:
## Changes

- Updated numerous version requirements: `llvmlite`, `numpy`, CUDA, etc
- Added `ipykernel`, `pygments`, and `pexpect` for optional tests
- Removed pin_compatible for `llvmlite` and `numpy` in favor of more explicit pinnings
- Updated tests (disable AVX512_SKX if detected with `numpy >=1.22` due to low accuracy)
- Updated `setuptools` upper limit
- Removed outdated patch

## Links

- GitHub: https://github.com/numba/numba/tree/0.56.3
- numba recipe: https://github.com/numba/numba/blob/0.56.3/buildscripts/condarecipe.local/meta.yaml
- conda-forge feedstock: https://github.com/conda-forge/numba-feedstock

## Notes

- `0.56.2` was skipped due to the `setuptools <60` requirement which could be problematic for some users. This upper limit was removed in `0.56.3`.
- `setuptools` remains a `run` dependency for now. For in-depth discussions see https://github.com/numba/numba/pull/8366 and https://github.com/numba/numba/pull/8475.
- The upper limit for `numpy` is set to `<1.23`, due to a recently-discovered CUDA compilation failure: https://github.com/numba/numba/issues/8529.
- GPU tests with CUDA were performed outside of Prefect, and passed for `linux-64` and `win-64`.

pseudoyim committed:
Updates for v0.56.2

pseudoyim committed:
Comment out patch

pseudoyim committed:
Add importlib_metadata for py<39

pseudoyim committed:
Remove patch; change incorporated here https://github.com/numba/numba/blob/0.56.2/numba/tests/test_parfors.py#L68-L70

pseudoyim committed:
Update llvmlite version to 0.39.x

pseudoyim committed:
Make URLs https

Commit:
Updated tests for ppc64le

Commit:
Added pip check to tests

Commit:
Updated tests to disable AVX512_SKX if detected with NumPy >= 1.22

Commit:
Updated sha256 and pinnings for numpy, cuda, openblas, tbb-devel

Commit:
Removed setuptools upper limit

Commit:
Fixed _openmp_mutex pinning for ppc64le

Commit:
Improved pinnings for llvmlite and numpy

Commit:
Added patch/m2-patch to build

Commit:
Adjust numpy upper limit down to <1.23 due to CUDA compilation failure

Commit:
Added setuptools upper limit

Commit:
Removed unneeded patches

Commit:
Added doc_source_url and license_url to about section

Commit:
Updated version to 0.56.3

Commit:
Removed setuptools <60 requirement

skupr-anaconda reviewed (approved):
LGTM

varlackc reviewed (commented):
`pip` is mentioned as a requirement
https://github.com/AnacondaRecipes/numba-feedstock/blob/720a60ef2e45579e16c0322161368b123b0bb984/recipe/meta.yaml#L37

We probably might need to add a `pip check` to the test section.

AndrewVallette commented:
> `pip` is mentioned as a requirement
> 
> https://github.com/AnacondaRecipes/numba-feedstock/blob/720a60ef2e45579e16c0322161368b123b0bb984/recipe/meta.yaml#L37
> 
> We probably might need to add a `pip check` to the test section.

`pip check` is performed in the test scripts. The `numba` team and `conda-forge` are doing it like this, so I opted to do the same.

ryanskeith reviewed (approved):
LGTM

cbouss commented on recipe/meta.yaml:
```diff
@@ -34,35 +31,42 @@ requirements:
     - setuptools
 
   run_constrained:
     - tbb  >=2021.0
```
Should tbb upper bound be aligned with the host definition? tbb-devel 2021.*,<2021.6

AndrewVallette replied:
I don't think `tbb` and `tbb-devel` have the same pinnings here. The documentation covers the requirements of TBB in depth here: https://numba.readthedocs.io/en/0.56.3/user/installing.html

- **tbb-devel** - provides TBB headers/libraries for compiling TBB support into Numba’s threading backend (`2021 <=` version `< 2021.6` required).
- **tbb** - provides the TBB runtime libraries used by Numba’s TBB threading backend (version `>= 2021` required).

**Edit:** Hmm, maybe a `2021.*` or `<2022` would be safer. Which do you think would be better?  I still think the `<2021.6` isn't quite right here -- the documentation conveys that it should work with any 2021.x version.

cbouss replied:
I think adding <2022 is the right thing to do. Especially, from what I heard, tbb uses 2021 and 2022 not as the year but as the major version.

AndrewVallette replied:
Thanks, I thought I remembered they do something a bit odd with the version here.

cbouss commented on recipe/meta.yaml:
```diff
@@ -34,35 +31,42 @@ requirements:
     - setuptools
     - wheel
-    - llvmlite 0.38.*
+    - llvmlite 0.39.*
```
Are pip and wheel required? I am not sure, but it likely doesn't hurt to leave them here.

AndrewVallette replied:
I'm not sure either, but I agree it's fine as-is.

cbouss commented on recipe/meta.yaml:
```diff
@@ -34,35 +31,42 @@ requirements:
+    - cuda-python >=11.6
 
 test:
   requires:
```
ipykernel is in numba's recipe:
https://github.com/numba/numba/blob/0.56.3/buildscripts/condarecipe.local/meta.yaml#L85

pygments and pexpect could also be added:
https://github.com/numba/numba/blob/0.56.3/buildscripts/incremental/setup_conda_environment.sh#L62

tbb also has an upper bound:
https://github.com/numba/numba/blob/0.56.3/buildscripts/incremental/setup_conda_environment.sh#L107

AndrewVallette replied:
Sure, `ipykernel`, `pygments`, and `pexpect` can be added, I can see they're used in some optional tests.

My `tbb` comments are here: https://github.com/AnacondaRecipes/numba-feedstock/pull/10#discussion_r1004014248

cbouss commented on recipe/meta.yaml:
```diff
@@ -34,35 +31,42 @@ requirements:
-    - {{ pin_compatible('llvmlite', max_pin='x.x') }}
-    - {{ pin_compatible('numpy') }}
+    - importlib_metadata       # [py<39]
+    - llvmlite >=0.39.*,<1
```
should be llvmlite >=0.39.*,<0.40
https://github.com/numba/numba/blob/0.56.3/setup.py#L28

AndrewVallette replied:
Thats right, thanks!

cbouss commented on recipe/meta.yaml:
```diff
@@ -34,35 +31,42 @@ requirements:
     - wheel
-    - llvmlite 0.38.*
+    - llvmlite 0.39.*
     - numpy
```
I think numpy should have the same restrictions in host and run:
numpy >=1.18,!=1.22.0,!=1.22.1,!=1.22.2,<1.23
See: https://github.com/numba/numba/blob/0.56.3/setup.py#L373

skupr-anaconda replied:
But if in `host` will be `numpy >=1.18,!=1.22.0,!=1.22.1,!=1.22.2,<1.23` then in `run` will be the **latest** numpy version and not the **oldest**.

cbouss replied:
Right, because of numpy's run export.
In that case numpy version should be set to the lowest supported: numpy 1.18

AndrewVallette replied:
To get this one working was quite a bit more complicated. I can see Serhii needed to do this before, so he might remember that `numpy` is tricky :)

I am using a `numpy_min_ver` variable to set the appropriate minimum version, depending on platform and Python version.

See the following commits: https://github.com/AnacondaRecipes/numba-feedstock/commit/d3b3890c29fe8fc959f7da3021e3ba94100d59b0 https://github.com/AnacondaRecipes/numba-feedstock/pull/10/commits/80f942b4a22347dcd839b3b54c65480ff310a286

cbouss reviewed (changes requested):
Mostly, I think the upper bounds should be updated in the run, run_constraint and test/requires sections for tbb and llvmlite

Commit:
Changed llvmlite upper run limit to <0.40

Commit:
Updated numpy pinnings in host

Commit:
Added ipykernel, pygments, and pexpect to test requirements for optional tests

Commit:
Changed upper limit for tbb to <2022

anaconda-pkg-build commented:
Linter check failed, with the following output:
The following problems have been found:

WARNING: /tmp/abs_830y91r65z/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends
ERROR: /tmp/abs_830y91r65z/clone/recipe/meta.yaml:41: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
ERROR: /tmp/abs_830y91r65z/clone/recipe/meta.yaml:44: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
Errors were found

anaconda-pkg-build commented:
Linter check failed, with the following output:
The following problems have been found:

WARNING: /tmp/abs_ce9xbatq0w/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends
ERROR: /tmp/abs_ce9xbatq0w/clone/recipe/meta.yaml:41: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
ERROR: /tmp/abs_ce9xbatq0w/clone/recipe/meta.yaml:44: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
Errors were found

anaconda-pkg-build commented:
Linter check failed, with the following output:
The following problems have been found:

WARNING: /tmp/abs_c45dnma9ov/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends
ERROR: /tmp/abs_c45dnma9ov/clone/recipe/meta.yaml:44: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
ERROR: /tmp/abs_c45dnma9ov/clone/recipe/meta.yaml:47: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
Errors were found

anaconda-pkg-build commented:
Linter check failed, with the following output:
The following problems have been found:

WARNING: /tmp/abs_38i0hlrwkt/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends
ERROR: /tmp/abs_38i0hlrwkt/clone/recipe/meta.yaml:44: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
ERROR: /tmp/abs_38i0hlrwkt/clone/recipe/meta.yaml:47: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
Errors were found

Commit:
Fixed numpy host pinnings

anaconda-pkg-build commented:
Linter check failed, with the following output:
The following problems have been found:

WARNING: /tmp/abs_5a4ybtzi5d/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends
ERROR: /tmp/abs_5a4ybtzi5d/clone/recipe/meta.yaml:45: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
ERROR: /tmp/abs_5a4ybtzi5d/clone/recipe/meta.yaml:48: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
Errors were found

AndrewVallette commented:
> Linter check failed, with the following output: The following problems have been found:
> 
> WARNING: /tmp/abs_5a4ybtzi5d/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends ERROR: /tmp/abs_5a4ybtzi5d/clone/recipe/meta.yaml:45: version_constraints_missing_whitespace: Packages and their version constraints must be space separated ERROR: /tmp/abs_5a4ybtzi5d/clone/recipe/meta.yaml:48: version_constraints_missing_whitespace: Packages and their version constraints must be space separated Errors were found

The linter is incorrect, and I have opened an issue: https://github.com/AnacondaRecipes/numba-feedstock/pull/10

Commit:
Use safer upper limit for numpy with python 3.10

anaconda-pkg-build commented:
Linter check failed, with the following output:
The following problems have been found:

WARNING: /tmp/abs_22ddqqu6l_/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends
ERROR: /tmp/abs_22ddqqu6l_/clone/recipe/meta.yaml:45: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
ERROR: /tmp/abs_22ddqqu6l_/clone/recipe/meta.yaml:48: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
Errors were found

anaconda-pkg-build commented:
Linter check failed, with the following output:
The following problems have been found:

WARNING: /tmp/abs_7f1drnh9c2/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends
ERROR: /tmp/abs_7f1drnh9c2/clone/recipe/meta.yaml:45: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
ERROR: /tmp/abs_7f1drnh9c2/clone/recipe/meta.yaml:48: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
Errors were found

AndrewVallette commented:
I believe all of the `numpy` pinnings have been resolved, as this is building for all platforms & Python versions now.

The only remaining issue is a DeprecationWarning in `pywin32` for one of the tests for `win-64` on `py310`. The warning is not an issue with `numba` itself.

The relevant part of the test failure is shown here:

```
INFO   FAIL: test_no_accidental_warnings (numba.tests.test_import.TestNumbaImport)
...
INFO     File "C:\b\abs_150hstvvni\croot\numba_1666764588839\_test_env\lib\site-packages\win32\lib\pywintypes.py", line 51, in __import_pywin32_system_module__
INFO       found = _win32sysloader.GetModuleFilename(filename)
INFO   DeprecationWarning: getargs: The 'u' format is deprecated. Use 'U' instead.

```

This seems to occur because our version of `pywin32` (302) is a little out of date. It's my opinion that this should not hold up the release, as it seems to be benign.

cbouss commented:
I agree. Based on this [link](https://docs.python.org/3/c-api/arg.html#strings-and-buffers) this will be removed in python 3.12.

Commit:
Added patch to ignore benign DeprecationWarning in test_import.py

pywin32 302 causes this warning to be emitted when pywintypes is imported:
"DeprecationWarning: getargs: The 'u' format is deprecated. Use 'U' instead."

This issue is only present on Windows with Python 3.10.

anaconda-pkg-build commented:
Linter check failed, with the following output:
The following problems have been found:

WARNING: /tmp/abs_93ldr97zwf/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends
ERROR: /tmp/abs_93ldr97zwf/clone/recipe/meta.yaml:49: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
ERROR: /tmp/abs_93ldr97zwf/clone/recipe/meta.yaml:52: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
Errors were found

AndrewVallette commented:
Re-triggering build due to test `win-64` test timeouts..

anaconda-pkg-build commented:
Linter check failed, with the following output:
The following problems have been found:

WARNING: /tmp/abs_912qypajtx/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends
ERROR: /tmp/abs_912qypajtx/clone/recipe/meta.yaml:49: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
ERROR: /tmp/abs_912qypajtx/clone/recipe/meta.yaml:52: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
Errors were found

AndrewVallette commented:
Prefect is having issues. Restarting again.

anaconda-pkg-build commented:
Linter check failed, with the following output:
The following problems have been found:

WARNING: /tmp/abs_b01zt032q5/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends
ERROR: /tmp/abs_b01zt032q5/clone/recipe/meta.yaml:49: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
ERROR: /tmp/abs_b01zt032q5/clone/recipe/meta.yaml:52: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
Errors were found

skupr-anaconda commented:
Only `win64` tests failed:
`AssertionError: Process timed out.`

cbouss reviewed (approved):
LGTM!

anaconda-pkg-build commented:
Linter check failed, with the following output:
The following problems have been found:

WARNING: /tmp/abs_b1n7eap52q/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends
ERROR: /tmp/abs_b1n7eap52q/clone/recipe/meta.yaml:49: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
ERROR: /tmp/abs_b1n7eap52q/clone/recipe/meta.yaml:52: version_constraints_missing_whitespace: Packages and their version constraints must be space separated
Errors were found
